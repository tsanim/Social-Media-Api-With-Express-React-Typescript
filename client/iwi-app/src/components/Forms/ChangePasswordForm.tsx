import React, { useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faLock, faKey } from '@fortawesome/free-solid-svg-icons';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import getFieldStyles from '../../utils/getFieldStyles';
import ChangePasswordFormProps from '../../interfaces/Components/Forms/ChangePasswordFormProps.interface';
import { ChangePassword } from '../../interfaces/Components/EditUserInformation/EditUserInfoProps.interface';

const validationSchema = Yup.object().shape({
    oldPassword: Yup.string()
        .required('Required'),
    newPassword: Yup.string()
        .matches(/^.*(?=.{6,})(?=.*[a-zA-Z])[a-zA-Z0-9]+$/, 'Password must contain at least one letter and at least one digit!')
        .min(8, 'Password must be at least 8 digits')
        .required('Required'),
});

function ChangePasswordForm({ changePasswordHandler, createSnackbar, errors: serverErrors, resetErrorsHandler }: ChangePasswordFormProps) {
    const handleSubmit = (form: ChangePassword) => {
        changePasswordHandler(form);

        createSnackbar({
            message: 'Password is changed!',
            timeout: 3000,
        });

        //clear all errors 
        resetErrorsHandler();
    }

    //clear all errors after component unmount, so the same errors not showing up on another component
    useEffect(() => {
        return () => {
            resetErrorsHandler();
        }
    }, [resetErrorsHandler]);

    return (
        <Formik
            onSubmit={handleSubmit}
            initialValues={{
                oldPassword: '',
                newPassword: ''
            }}
            validationSchema={validationSchema}
        >
            {
                ({ errors }: any) => {
                    return (
                        <Form className="editForm">
                            {
                                (serverErrors.length > 0 && (<p className="error">{serverErrors[serverErrors.length - 1].message}</p>))
                            }
                            <span>
                                <FontAwesomeIcon icon={faKey} />
                                <Field
                                    type="password"
                                    name="oldPassword"
                                    placeholder="Old password..."
                                    style={getFieldStyles(errors, 'oldPassword')}
                                />
                                <ErrorMessage className="error-message" name="oldPassword" component="div" />
                            </span>
                            <span>
                                <FontAwesomeIcon icon={faLock} />
                                <Field
                                    type="password"
                                    name="newPassword"
                                    placeholder="New password..."
                                    style={getFieldStyles(errors, 'newPassword')}
                                />
                                <ErrorMessage className="error-message" name="newPassword" component="div" />
                            </span>
                            <input type="submit" value="CHANGE" />
                        </Form>
                    )
                }
            }
        </Formik>

    )
}

ChangePasswordForm.propTypes = {
    changePasswordHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
    errors: PropTypes.array,
    resetErrorsHandler: PropTypes.func
}

export default wrapComponent(ChangePasswordForm);