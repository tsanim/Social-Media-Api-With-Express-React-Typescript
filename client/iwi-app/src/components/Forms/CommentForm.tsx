import React from 'react';
import useForms from '../../hooks/useForms';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import CommentFormProps from '../../interfaces/Components/Forms/CommentFormProps.interface';

export function CommentForm({ postId, makeCommentHandler, createSnackbar }: CommentFormProps) {
    const { handleSubmit, handleChangeInput, inputs } = useForms(() => {
        //check for empty text and send message to user
        if (inputs.text === '' || !inputs.text) {
            createSnackbar({
                message: 'You can not make comment without text!',
                timeout: 3000,
            });
        } else {
            makeCommentHandler({ ...inputs, postId })
        }
    });

    return (
        <form className="commentForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write a comment..."
                name="text"
                id="comment"
                value={inputs.text || ''}
                onChange={handleChangeInput}
            />
            <input disabled={inputs.text === '' || !inputs.text} type="submit" value="Comment" />
        </form>
    )
}

CommentForm.propTypes = {
    postId: PropTypes.string,
    makeCommentHandler: PropTypes.func,
    createSnackbar: PropTypes.func
}

export default wrapComponent(CommentForm);
