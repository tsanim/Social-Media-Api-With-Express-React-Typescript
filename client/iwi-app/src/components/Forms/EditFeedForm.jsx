import React from 'react';
import useForms from '../../hooks/useForms';
import PropTypes from 'prop-types';

function EditFeedForm({ feedId, text, editUserPostHandler, handleShowEditForm }) {
    const { handleSubmit, handleChangeInput, inputs } = useForms((e) => {
        editUserPostHandler(inputs, feedId);
        handleShowEditForm();
    });

    return (
        <form id="editFeedForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write something..."
                name="text"
                id="feed"
                value={inputs.text || text}
                onChange={handleChangeInput}
            />
            <input type="submit" value="SAVE" />
            <button onClick={handleShowEditForm}>Undo</button>
        </form>
    )
}

EditFeedForm.propTypes = {
    feedId: PropTypes.string,
    text: PropTypes.string,
    editUserPostHandler: PropTypes.func,
    handleShowEditForm: PropTypes.func,
}

export default EditFeedForm;