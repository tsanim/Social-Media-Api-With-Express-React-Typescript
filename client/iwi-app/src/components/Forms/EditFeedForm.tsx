import React from 'react';
import useForms from '../../hooks/useForms';
import PropTypes from 'prop-types';
import EditFeedFormProps from '../../interfaces/Components/Forms/EditFeedFormProps.interface';

function EditFeedForm({ feedId, text, editUserPostHandler, handleShowEditForm }: EditFeedFormProps) {
    const { handleSubmit, handleChangeInput, inputs } = useForms(() => {
        editUserPostHandler(inputs, feedId);
        handleShowEditForm();
    });

    return (
        <form id="editFeedForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write something..."
                name="text"
                id="feed"
                value={inputs.text || text}
                onChange={handleChangeInput}
            />
            <input type="submit" value="SAVE" />
            <button onClick={handleShowEditForm}>Undo</button>
        </form>
    )
}

EditFeedForm.propTypes = {
    feedId: PropTypes.string,
    text: PropTypes.string,
    editUserPostHandler: PropTypes.func,
    handleShowEditForm: PropTypes.func,
}

export default EditFeedForm;