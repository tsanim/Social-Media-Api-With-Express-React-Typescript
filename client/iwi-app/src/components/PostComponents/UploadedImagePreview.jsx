import React from 'react';
import PropTypes from 'prop-types';

function UploadedImagePreview(props) {
    return (
        <div className="uploadedImgPreview">
            {
                props.imagePreviewUrl !== '' ? <React.Fragment>
                    <img id="previewImg" src={props.imagePreviewUrl} alt="uploadedImg" />
                    <div onClick={props.handlePreviewImgClose} className="overlay"><span>x</span></div>
                </React.Fragment> : null
            }
        </div>
    )
}

UploadedImagePreview.propTypes = {
    imagePreviewUrl: PropTypes.string,
    handlePreviewImgClose: PropTypes.func
}

export default UploadedImagePreview;