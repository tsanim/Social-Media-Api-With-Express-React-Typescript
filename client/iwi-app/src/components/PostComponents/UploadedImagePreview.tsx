import React from 'react';
import PropTypes from 'prop-types';
import UploadedImagePreviewProps from '../../interfaces/Components/PostComponents/UploadedImagePreviewProps.interface';

function UploadedImagePreview(props: UploadedImagePreviewProps) {
    return (
        <div className="uploadedImgPreview">
            {
                props.imagePreviewUrl !== '' ? <React.Fragment>
                    <img id="previewImg" src={props.imagePreviewUrl} alt="uploadedImg" />
                    <div onClick={props.handlePreviewImgClose} className="overlay"><span>x</span></div>
                </React.Fragment> : null
            }
        </div>
    )
}

UploadedImagePreview.propTypes = {
    imagePreviewUrl: PropTypes.string,
    handlePreviewImgClose: PropTypes.func
}

export default UploadedImagePreview;