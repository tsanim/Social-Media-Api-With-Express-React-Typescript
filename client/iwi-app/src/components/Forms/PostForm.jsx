import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import isImage from '../../utils/isImage';
import UploadedImagePreview from '../PostComponents/UploadedImagePreview';

let fd = new FormData();

export class PostForm extends React.Component {
    state = {
        text: '',
        imagePreviewUrl: '',
    }

    handleChangeInput = (e) => {
        e.persist();

        const files = e.target.files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = e.target.files[0];

            if (!isImage(file.type)) {
                this.props.createSnackbar({
                    message: 'You should upload only images!',
                    timeout: 3000,
                });
                return;
            }

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState((oldState) => ({
                    imagePreviewUrl: reader.result,
                }));
            }

            reader.readAsDataURL(file);
        }

        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    handleSubmit = (e) => {
        e.preventDefault();

        //check for empty text and send message to user or emtpy file 
        if (this.state.text === '' && !fd.get('postImg')) {
            //using createSnackbar for displaying warning message to user 
            this.props.createSnackbar({
                message: 'You can not upload post without image or text!',
                timeout: 3000,
            });
        } else {
            fd.append('text', this.state.text);
            this.props.uploadHandler(fd)
        }

        //reset state
        fd = new FormData();
        this.setState({ text: '', imagePreviewUrl: '' });
    }

    handlePreviewImgClose = (e) => {
        fd.delete('postImg');
        this.setState({ imagePreviewUrl: '' })
    }

    render() {
        const { text, imagePreviewUrl } = this.state;

        return (
            <div className="postForm">
                <form id="postForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                    <textarea
                        name="text"
                        id="postText"
                        placeholder="Write something to your friends..."
                        value={text}
                        onChange={this.handleChangeInput}
                    />
                    <div className="inline-form-items">
                        <div className="uploadImage">
                            <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Image</label>
                            <input
                                type="file"
                                id="file"
                                name="postImg"
                                onChange={this.handleChangeInput}
                            />
                        </div>
                        <input disabled={!(fd.keys() > 0 || (text !== '' || imagePreviewUrl !== ''))} type="submit" value="Post" />
                    </div>
                </form>

                <UploadedImagePreview
                    imagePreviewUrl={imagePreviewUrl}
                    handlePreviewImgClose={this.handlePreviewImgClose}
                />
            </div>
        )
    }
}



PostForm.propTypes = {
    uploadHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
}

export default wrapComponent(PostForm);