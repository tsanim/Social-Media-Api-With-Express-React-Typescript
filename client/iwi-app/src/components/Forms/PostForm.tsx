import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import isImage from '../../utils/isImage';
import UploadedImagePreview from '../PostComponents/UploadedImagePreview';
import PostFormState from '../../interfaces/Components/Forms/PostFormState.interface';
import PostFormProps from '../../interfaces/Components/Forms/PostFomProps.interface';

let fd = new FormData();

export class PostForm extends React.Component<PostFormProps, PostFormState> {
    state = {
        text: '',
        imagePreviewUrl: '',
    }

    handleChangeInput = (e: React.ChangeEvent<HTMLTextAreaElement> | React.ChangeEvent<HTMLInputElement>) => {
        e.persist();

        const files = (e.target as HTMLInputElement).files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = (e.target as HTMLInputElement).files[0];

            if (!isImage(file.type)) {
                this.props.createSnackbar({
                    message: 'You should upload only images!',
                    timeout: 3000,
                });
                return;
            }

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState((oldState) => ({
                    imagePreviewUrl: reader.result as string,
                }));
            }

            reader.readAsDataURL(file);
        }

        this.setState({
            text: e.target.value,
        });
    }

    handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        //check for empty text and send message to user or emtpy file 
        if (this.state.text === '' && !fd.get('postImg')) {
            //using createSnackbar for displaying warning message to user 
            this.props.createSnackbar({
                message: 'You can not upload post without image or text!',
                timeout: 3000,
            });
        } else {
            fd.append('text', this.state.text);
            this.props.uploadHandler(fd);
        }

        //reset state
        fd = new FormData();
        this.setState({ text: '', imagePreviewUrl: '' });
    }

    handlePreviewImgClose = () => {
        fd.delete('postImg');
        this.setState({ imagePreviewUrl: '' })
    }

    render() {
        const { text, imagePreviewUrl } = this.state;

        return (
            <div className="postForm">
                <form id="postForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                    <textarea
                        name="text"
                        id="postText"
                        placeholder="Write something to your friends..."
                        value={text}
                        onChange={this.handleChangeInput}
                    />
                    <div className="inline-form-items">
                        <div className="uploadImage">
                            <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Image</label>
                            <input
                                type="file"
                                id="file"
                                name="postImg"
                                onChange={this.handleChangeInput}
                            />
                        </div>
                        <input disabled={!(fd.keys() || (text !== '' || imagePreviewUrl !== ''))} type="submit" value="Post" />
                    </div>
                </form>

                <UploadedImagePreview
                    imagePreviewUrl={imagePreviewUrl}
                    handlePreviewImgClose={this.handlePreviewImgClose}
                />
            </div>
        )
    }
}

// PostForm.propTypes = {
//     uploadHandler: PropTypes.func,
//     createSnackbar: PropTypes.func,
// }

export default wrapComponent(PostForm);