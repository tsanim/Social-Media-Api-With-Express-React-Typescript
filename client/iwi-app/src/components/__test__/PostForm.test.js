import React from 'react';
import { shallow, mount } from 'enzyme';

import { PostForm } from '../Forms/PostForm';
import UploadedImagePreview from '../PostComponents/UploadedImagePreview';
import { MemoryRouter } from 'react-router-dom';

function postFormComponentSetup() {
    const props = {
        uploadHandler: jest.fn(),
        createSnackbar: jest.fn(),
    }

    const state = {
        text: '',
        imagePreviewUrl: '',
    }

    const enzymeWrapper = shallow(<PostForm {...props} />)

    enzymeWrapper.setState(state);

    return {
        props,
        enzymeWrapper,
        state
    }
}


describe('PostForm Component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render container div tag with class "postForm"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('div.postForm')).toHaveLength(1);
    });

    it('should render container form tag with id "postForm"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('form#postForm')).toHaveLength(1);
    });

    it('should render textarea tag', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('textarea')).toHaveLength(1);
    });

    it('should render div tag with class "inline-form-items"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('div.inline-form-items')).toHaveLength(1);
    });

    it('should render div tag with class "uploadImage"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('div.uploadImage')).toHaveLength(1);
    });

    it('should render input with type "file"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('input[type="file"]')).toHaveLength(1);
    });

    it('should render img tag with class "icon"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('img.icon')).toHaveLength(1);
    });

    it('should render input with type "file" with onChange prop', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('input[type="file"]').prop('onChange')).toBeDefined();
    });

    it('should render input with type "submit"', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('input[type="submit"]')).toHaveLength(1);
    });

    it('should render input with type "submit" to be disabled', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('input[type="submit"]').prop('disabled')).toEqual(true);
    });

    it('should render form tag with encType prop', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        expect(enzymeWrapper.find('form').prop('encType')).toEqual('multipart/form-data');
    });

    it('should render UploadedImagePreview Compoenent if imagePreviewUrl is not empty string', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        enzymeWrapper.setState({ imagePreviewUrl: '123' });

        expect(enzymeWrapper.find(UploadedImagePreview)).toHaveLength(1);
    });

    it('should test if uploadHandler() is called when form is submitted and has post text or post image', () => {
        const { enzymeWrapper, props } = postFormComponentSetup();

        const mockedEvent = {
            target: { value: 'hello', name: 'text' },
            persist: jest.fn()
        };

        enzymeWrapper.find('textarea').simulate('change', mockedEvent);
        enzymeWrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(props.uploadHandler).toHaveBeenCalledTimes(1);
    });

    it('should test if createSnackbar() is called when form is submitted and has not post text or post image', () => {
        const { enzymeWrapper, props } = postFormComponentSetup();

        enzymeWrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(props.createSnackbar).toHaveBeenCalledTimes(1);
    });

    it('should test if user text is echoed', () => {
        const { enzymeWrapper } = postFormComponentSetup();

        const mockedEvent = {
            target: { value: 'hello', name: 'text' },
            persist: jest.fn()
        };

        enzymeWrapper.find('textarea').simulate('change', mockedEvent);

        expect(enzymeWrapper.find('textarea').props().value).toEqual('hello');
    });

    it('should test if input type file onChange handler call createSnackbar when file type is not image', () => {
        const { enzymeWrapper, props } = postFormComponentSetup();

        const file = new Blob([
            JSON.stringify({})
        ], { type: 'text/plain' });

        enzymeWrapper.find('input[type="file"]').simulate('change', {
            preventDefault: () => { },
            persist: () => { },
            target: {
                files: [file],
                name: 'postImg'
            }
        });

        expect(props.createSnackbar).toHaveBeenCalledTimes(1);
    });

    it('should test if file input onChange prop is called and echoed', (done) => {
        const { props } = postFormComponentSetup();
        const wrapper = mount(<MemoryRouter>
            <PostForm {...props} />
        </MemoryRouter>)
        const file = new Blob([
            JSON.stringify({})
        ], { type: 'image/jpeg' });

        const reader = new FileReader();

        wrapper.find('input[type="file"]').simulate('change', {
            preventDefault: jest.fn(),
            persist: jest.fn(),
            target: {
                files: [file],
                name: 'postImg'
            }
        });

        reader.onload = () => {
            wrapper.setState({ imagePreviewUrl: reader.result });
            expect(wrapper.find('img#previewImg').prop('src')).toEqual(reader.result);
            done();
        }

        reader.readAsDataURL(file);
        done();
    });
});