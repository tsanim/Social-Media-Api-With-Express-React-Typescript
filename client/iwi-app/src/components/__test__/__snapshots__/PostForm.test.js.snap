// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm Component should render initial layout 1`] = `
Array [
  <div
    className="postForm"
  >
    <form
      encType="multipart/form-data"
      id="postForm"
      onSubmit={[Function]}
    >
      <textarea
        id="postText"
        name="text"
        onChange={[Function]}
        placeholder="Write something to your friends..."
        value=""
      />
      <div
        className="inline-form-items"
      >
        <div
          className="uploadImage"
        >
          <label
            htmlFor="file"
          >
            <img
              alt="uploadImg"
              className="icon"
              src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png"
            />
            Image
          </label>
          <input
            id="file"
            name="postImg"
            onChange={[Function]}
            type="file"
          />
        </div>
        <input
          disabled={true}
          type="submit"
          value="Post"
        />
      </div>
    </form>
    <UploadedImagePreview
      handlePreviewImgClose={[Function]}
      imagePreviewUrl=""
    />
  </div>,
]
`;
